### ISSUE TITLE
Cannot connect to rabbitmq_stomp

### ISSUE THREAD

**Issue by nkvoll** [mozes/stompest#2]
I installed the stomp plugin for rabbitmq (rabbitmq_stomp.ez and amqp_client.ez) as per http://www.rabbitmq.com/stomp.html.

```
>>> from stompest.simple import Stomp
>>> QUEUE = '/queue/simpleTest'
>>> stomp = Stomp('localhost', 61613)
>>> stomp.connect()
```

After this, the python shell hangs indefinitely.

Testing against rabbitmq with telnet works as expected:

```
$ telnet localhost 61613
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
CONNECT

^@
CONNECTED
session:session-9Z1PNAe1n3GvJFeAApbMPg==
heart-beat:0,0
version:1.0

DISCONNECT

^@
Connection closed by foreign host.
```

While I was expecting some compatibility issues ("your milage may wary"), I was hoping basic functionality would be there :)

**Comment by ghost** [comment]
Do you also see this behavior with the twisted client?

**Comment by nkvoll** [comment]
Yes, I get the same behavior with the twisted client (from the readme example).

```
$ piped stomp_producer.py 
DEBUG:stompest.async:Sending connect command
```

Using rabbitmq-2.6.1.

**Comment by nkvoll** [comment]
I CTRL-C'ed the process after a while and got this:

```
$ python stomp_producer.py 
DEBUG:stompest.async:Sending connect command
^CDEBUG:stompest.async:Disconnected: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionLost'>: Connection to the other side was lost in a non-clean fashion: Connection lost.
]
ERROR:stompest.async:Connection lost with outstanding connectedDeferred
DEBUG:stompest.async:Calling connectedDeferred errback: Unexpected connection loss
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: stompest.error.StompError: Unexpected connection loss
$
```

**Comment by theduderog** [comment]
I've been trying to run the stomp plugin on Ubuntu 10.04 but so far haven't figured out how to get past this error.

starting STOMP Adapter (binding to {[61613],[]})  ...{"Kernel pid terminated",application_controller,"{application_start_failure,rabbitmq_stomp,{bad_return,{{rabbit_stomp,start,[normal,[]]},{'EXIT',{{badmatch,{error,{undef,[{rabbit_networking,check_tcp_listener_address,[rabbit_stomp_listener_sup,61613]},{rabbit_stomp_sup,'-listener_specs/3-lc$^0/1-0-',3},{rabbit_stomp_sup,init,1},{supervisor,init,1},{gen_server,init_it,6},{proc_lib,init_p_do_apply,3}]}}},[{rabbit_stomp,start,2},{application_master,start_it_old,4}]}}}}}"}
Erlang has closed
Error: {node_start_failed,normal}

Crash dump was written to: erl_crash.dump
Kernel pid terminated (application_controller) ({application_start_failure,rabbitmq_stomp,{bad_return,{{rabbit_stomp,start,[normal,[]]},{'EXIT',{{badmatch,{error,{undef,[{rabbit_networking,check_tcp

**Comment by theduderog** [comment]
Hmm...finally got Ubuntu 11.04 running with rabbitmq-stomp plugin but it abruptly closes the connection when I try to connect with telnet.

$ telnet localhost 61613
Trying ::1...
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
CONNECT
login: guest
password: guest

Connection closed by foreign host.

This is the behavior I see from the python client:

> > > from stompest.simple import Stomp
> > > QUEUE = '/queue/simpleTest'
> > > stomp = Stomp('localhost', 61613)
> > > stomp.connect()
> > > Traceback (most recent call last):
> > >   File "<stdin>", line 1, in <module>
> > >   File "/media/psf/Work/stompest/stompest/simple.py", line 39, in connect
> > >     frame = self.receiveFrame()
> > >   File "/media/psf/Work/stompest/stompest/simple.py", line 80, in receiveFrame
> > >     parser.processLine(self.sfile.readline()[:-1])
> > >   File "/media/psf/Work/stompest/stompest/parser.py", line 68, in processLine
> > >     self.states[self.state](line)
> > >   File "/media/psf/Work/stompest/stompest/parser.py", line 78, in parseCommandLine
> > >     raise StompFrameError("Empty stomp command line: %s" % line)
> > > stompest.error.StompFrameError: Empty stomp command line:

**Comment by nkvoll** [comment]
Telnetting, netcating and other clients works for me.

Have you tried using netcat by the way? http://www.rabbitmq.com/stomp.html explains it (see the section "Testing the Adapter")

(output from when I tried your telnet transcript above follows):

```
$ telnet localhost 61613
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
CONNECT
login: guest
password: guest

^@
CONNECTED
session:session-EC+5a9NPimyK3Wm/d99cBQ==
heart-beat:0,0
version:1.0
```

**Comment by nkvoll** [comment]
I've researched a little, and it does not seem like rabbitmq adds a trailing \n to the \x00 bytes, which stompest seem to assume is there, so it never parses the frames as received, since socket.readline() and equivalent methods only returns on \n.

**Comment by ghost** [comment]
Yeah, I came to the same conclusion. Unfortunately, according to the STOMP protocol it seems that the newline-after-nul is optional. It pretty much kills the approach in both the simple and twisted clients, since they both expect all lines to be terminated with newlines.

simple is kind of trivial to fix, its just a matter of reading chunks from the socket and breaking on a nulbyte (something similar to http://code.google.com/p/stomppy/source/browse/stomp/connect.py, around line 623). I'm pretty new to twisted so unsure what alternative there is to LineReceiver.

**Comment by nkvoll** [comment]
From the STOMP specification:

"The frame starts with a command string terminated by a newline. Following the command are one or more header entries in <key>:<value> format. Each header entry is terminated by a newline. A blank line indicates the end of the headers and the beginning of the body. The body is then followed by the null byte (0x00). The examples in this document will use ^@, control-@ in ASCII, to represent the null byte. The null byte can be optionally followed by multiple newlines. "

As I understand it, this means the newlines after the null-byte should be ignored. I'm writing a fix for both the simple and the twisted "backends" which should be ready in a couple of minutes :)

**Comment by nkvoll** [comment]
Can you take a look at https://github.com/mozes/stompest/pull/3 ?

**Comment by nkvoll** [comment]
Fixed by #3. Thanks for the quick responses!
